### ISSUE TITLE
Polymorphic Through

### ISSUE THREAD

**Issue by connected-pjeri** [RobinBuschmann/sequelize-typescript#160]
Using: `"sequelize-typescript": "0.4.0-beta"`

I'm having some trouble figuring out how to do polymorphic model definitions so that the queries work as expected. I'm trying to do a **Subscription** table eg:

```sql
Table: Subscription
integer: userId
integer: targertId
string: targetType  -> One of USER / POST / ETC
```

I'm focusing on getting **USER** to work first. I want to be able to get all the users a particular user has subscribed to. So I have something like this:

```typescript
class User extends Model<User> {
...
    @HasMany(
    () => Subscription,
    {
      foreignKey: "targetId",
      scope: {
        targetType: "USER",
      },
    }
  )
  usersSubscribedTo: IUser[];
...
}
```

```typescript
class Subscription extends Model<Subscription> {
...
  @ForeignKey(() => User)
  @AllowNull(false)
  @Column
  userId: number;

  @BelongsTo (
    () => User,
    "userId",
  )
  user: IUser;

  @AllowNull(false)
  @Column
  targetId: number;

  @AllowNull(false)
  @Column
  targetType: string;

  @BelongsTo (
    () => User,
    {
      foreignKey: "targetId",
      constraints: false,
      scope: { targetType: "USER" },
    },
  )
  userSubscribedTo: User;
...
}
```

This looks kind of ok to me, but with this I will get:
```
error TS2345: Argument of type '{ foreignKey: string; constraints: false; scope: { targetType: string; }; }' is not assignable to parameter of type 'AssociationOptionsBelongsTo | undefined'.
  Object literal may only specify known properties, and 'scope' does not exist in type 'AssociationOptionsBelongsTo | undefined'.
```

**Comment by connected-pjeri** [comment]
**Update**
How could I achieve it if I treat Subscription as a through table:

```typescript
  @HasMany({
    () => User,
    through: () => Subscription,
    constraints: false,
    scope: { targetType: "USER" },
  }, "targetId")
  usersSubscribedTo: IUser[];
```

**Comment by RobinBuschmann** [comment]
So you want to implement a many-to-many relation, where the subscription table is the through table?

**Comment by connected-pjeri** [comment]
Correct. So that ultimately I can query things like:
```
user.subscribedUsers
user.subscribedPosts
post.subscribedUsers
```

**Comment by RobinBuschmann** [comment]
Did you tried something like:

```typescript
@Table
export class User extends Model<User> {

    @BelongsToMany(() => User, () => Subscription, 'subscriberId', 'userId')
    usersSubscribedTo: User[];
}

@Table
export class Subscription extends Model<Subscription> {

    @PrimaryKey
    @ForeignKey(() => User)
    @Column
    subscriberId: number;

    @PrimaryKey
    @ForeignKey(() => User)
    @Column
    userId: number;
}
```

**Comment by connected-pjeri** [comment]
If you remove the `subscriberType` column then that's not polymorphic.

**Comment by RobinBuschmann** [comment]
Ah, sry, my fault. This would be the proper way then:

```typescript
@Table
export class User extends Model<User> {

    @Column
    name: string;

    @BelongsToMany(() => User, {
        through: () => Subscription,
        foreignKey:  'subscriberId',
        otherKey: 'targetId',
        constraints: false,
        foreignKeyConstraint: true,
        scope: {
            targetType: 'user'
        }
    })
    usersSubscribedTo: User[];

    @BelongsToMany(() => Post, {
        through: () => Subscription,
        foreignKey:  'subscriberId',
        otherKey: 'targetId',
        constraints: false,
        foreignKeyConstraint: true,
        scope: {
            targetType: 'post'
        }
    })
    postsSubscribedTo: Post[];
}

@Table
export class Subscription extends Model<Subscription> {

    @PrimaryKey
    @ForeignKey(() => User)
    @Column
    subscriberId: number;

    @PrimaryKey
    @Column
    targetId: number;

    @Column
    targetType: string;
}
```

But when querying the data, the created sql string doesn't seem to work unfortunately. I will look into this soon.

**Comment by RobinBuschmann** [comment]
What I can already say is, that querying data with this model configuration doesn't work with pure sequelize, too.

**Comment by RobinBuschmann** [comment]
OK, it is possible. But therefor it is necessary to set through options to the belongsToMany options. See #122

**Comment by connected-pjeri** [comment]
I see. I appreciate you looking into this in such a short time @RobinBuschmann :)

**Comment by RobinBuschmann** [comment]
`ThroughOptions` are now available since `0.6.0-beta.3` (latest release: `0.6.0-beta.5`).
